File: c/Observer/Display.c
#include "Display.h"

void CurrentConditionsDisplayNew(CurrentConditionsDisplay *this, WeatherStation *subject)
{
    this->update = CCDUpdate;
    this->isSameObserver = CCDIsSameObserver;
    this->display = CCDDisplay;

    subject->registerObserver(subject, (IObserver *)this);
}
void CCDUpdate(CurrentConditionsDisplay *this, void *data)
{
    this->data = *(WeatherData *)data;
    this->display(this);
}
bool CCDIsSameObserver(CurrentConditionsDisplay *this, IObserver *observer)
{
    return memcmp(this, observer, sizeof(CurrentConditionsDisplay)) == 0;
}
void CCDDisplay(CurrentConditionsDisplay *this)
{
    printf("Current Condition :\n");
    printf("temperature : %0.2f\n", this->data.temperature);
    printf("humidity : %0.2f\n", this->data.humidity);
    printf("pressure : %0.2f\n", this->data.pressure);
}
void StatisticsDisplayNew(StatisticsDisplay *this, WeatherStation *subject)
{
    this->data_count = 0;
    this->update = SDUpdate;
    this->isSameObserver = SDIsSameObserver;
    this->display = SDDisplay;
    this->data.temperature = 0.0;
    this->data.humidity = 0.0;
    this->data.pressure = 0.0;

    subject->registerObserver(subject, (IObserver *)this);
}
void SDUpdate(StatisticsDisplay *this, void *data)
{
    AddWeatherData(&this->data, (WeatherData *)data);
    this->data_count++;
    this->display(this);
}
bool SDIsSameObserver(StatisticsDisplay *this, IObserver *observer)
{
    return memcmp(this, observer, sizeof(StatisticsDisplay)) == 0;
}
void SDDisplay(StatisticsDisplay *this)
{
    printf("Statistics Data :\n");
    printf("avg temperature : %0.2f\n", this->data.temperature / this->data_count);
    printf("avg humidity : %0.2f\n", this->data.humidity / this->data_count);
    printf("avg pressure : %0.2f\n", this->data.pressure / this->data_count);
}

File: python/Observer/WeatherStation/Display.py
from .interface.Observer import Observer
from .interface.Subject import Subject
from .interface.DisplayElement import DisplayElement
from .interface.Data import Data


class CurrentConditionsDisplay(Observer, DisplayElement):
    def __init__(self, weather_data: Subject):
        self.weather_data: Subject = weather_data
        self.object: Data = Data()

        self.weather_data.registerObserver(self)

    def update(self, object: Data):
        self.object = object
        self.display()

    def display(self):
        print("Current Condition :")
        print(f'temperature : {self.object.temperature}')
        print(f'humidity : {self.object.humidity}')
        print(f'pressure : {self.object.pressure}')


class StatisticsDisplay(Observer, DisplayElement):
    def __init__(self, weather_data: Subject):
        self.weather_data: Subject = weather_data
        self.object: Data = Data()
        self.count: int = 0

        self.weather_data.registerObserver(self)

    def update(self, object: Data):
        def caculate(old_data: float, old_count: float, new_data: float) -> float:
            return (old_data*old_count+new_data)/(old_count+1)

        self.object.temperature = caculate(
            self.object.temperature, self.count, object.temperature)
        self.object.humidity = caculate(
            self.object.humidity, self.count, object.humidity)
        self.object.pressure = caculate(
            self.object.pressure, self.count, object.pressure)
        self.count += 1
        self.display()

    def display(self):
        print("Statistics Data :")
        print(f'avg temperature : {self.object.temperature:0.2f}')
        print(f'avg humidity : {self.object.humidity:0.2f}')
        print(f'avg pressure : {self.object.pressure:0.2f}')
